transport-native-epoll/src/main/java/io/netty/channel/epoll/TunTapChannel.java

```diff
@@ -0,0 +1,298 @@
+        // (see doBind() below).  If the file descriptor is added to the epoll set before the
+        // network interface is allocated, epoll will never wake on incoming packets.  This
+        // appears to be an idiosyncrasy of the Linux kernel.
+    }
+
+    @Override
+    protected void doBind(SocketAddress localAddress) throws Exception {
+        if (isActive()) {
+            throw new IllegalStateException("TunTapChannel already bound to interface");
+        }
+
+        TunTapAddress tunTapAddr = (TunTapAddress) localAddress;
+
+        String ifName = tunTapAddr.interfaceName();
+        boolean isTapDevice = tunTapAddr.isTapAddress();
+
+        // Allocate/open the specified network interface name.
+        fd().allocTunTapInterface(ifName, isTapDevice);
+
+        _localAddress = tunTapAddr;
+        _isTapChannel = isTapDevice;
+
+        // Add the TUN/TAP file descriptor to the epoll set.
+        // NOTE: this must happen *after* the call to allocInterface() above (see
+        // doRegister() for further details).
+        EpollEventLoop loop = (EpollEventLoop) eventLoop();
+        loop.add(this);
+
+        // Mark the channel as active.
+        active = true;
+    }
+
+    @Override
+    protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+        for (;;) {
+            // Get the next packet to be written.  Bail if no more packets.
+            TunTapPacket packet = (TunTapPacket) in.current();
+            if (packet == null) {
+                break;
+            }
+
+            ByteBuf packetData = packet.packetData();
+
+            try {
+                // If the buffer has an accessible memory address, write the packet data
+                // directly from buffer memory. Otherwise, get the packet data as an NIO
+                // buffer and write the data from that.
+                if (packetData.hasMemoryAddress()) {
+                    long memoryAddress = packetData.memoryAddress();
+                    int readerIndex = packetData.readerIndex();
+                    int len = packetData.readableBytes();
+                    fd().writeTunTapPacketAddress(packet.protocol(), memoryAddress, readerIndex, len);
+                } else {
+                    ByteBuffer nioBuf = packetData.nioBuffer();
+                    int pos = nioBuf.position();
+                    int limit = nioBuf.limit();
+                    fd().writeTunTapPacket(packet.protocol(), nioBuf, pos, limit - pos);
+                }
+
+                // Drop the packet from the outbound queue.
+                in.remove();
+            } catch (Exception e) {
+                in.remove(e);
+            }
+        }
+    }
+
+    @Override
+    protected AbstractEpollUnsafe newUnsafe() {
+        return new TunTapChannelUnsafe();
+    }
+
+    final class TunTapChannelUnsafe extends AbstractEpollUnsafe {
+
+        private final List<TunTapPacket> _rcvdPackets = new ArrayList<TunTapPacket>();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Rename this field "_rcvdPackets" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS00116)